### PR TITLE
feat: show current email during password reset and auto-populate text-box after successful completion

### PR DESCRIPTION
**Pull request by namit-chandwani**
## Linked Issues
- Closes https://github.com/sourcegraph/sourcegraph/issues/38348

## Motivation and Context:
 
- Improves the UX of the password reset flow

## Changes Made:
 
- Made changes to the following 2 flows:
	- On the new password entry screen:
		- Added a text which displays the email of the account for which the password change request has been raised
		- Added a back button to allow the users to go back to the previous email entry screen if the account they want to reset the password for is different
	- On the sign-in screen which comes after successful password reset request completion:
		- Made changes to auto-populate the email text-box with the email linked to the account on which the password reset request was completed recently

## Type of change:

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Refactoring (altering code without changing its external behaviour)
- [ ] Documentation change
- [ ] Other

## Checklist:

- [x] Development completed
- [ ] Comments added to code (wherever necessary)
- [ ] Documentation updated (if applicable)
- [x] Tested changes locally

## Follow-up tasks (if any):

- None

## Test Plan

 - Setup the codebase locally along with configuring a custom SMTP to enable the email delivery functionality with the help of this documentation: https://docs.sourcegraph.com/admin/config/email#configuring-sourcegraph-to-send-email-using-another-provider

 - Tested the entire flow locally end-to-end.
 - Screen recording of the password reset screen where the current email ID is added along with a back button:
	 	
https://github.com/sourcegraph/sourcegraph/assets/51479159/a79fc338-ace0-4281-86d2-de7cc68eae20
	 
 - Screen recording of the sign-in screen after password reset is successfully done where the email ID is auto-populated in the text-box:
	 	
https://github.com/sourcegraph/sourcegraph/assets/51479159/be7db65d-9421-4621-a1e9-a04a546b9757

    
## Additional Comments
- Please let me know if I need to make any further design changes from the frontend side or any API contract related changes from the backend side

### CODE REVIEW

**Comment by bobheadxi** [comment]
Got a request for reviews for this contribution via https://github.com/sourcegraph/sourcegraph/issues/38348#issuecomment-2121019664, tagged the appropriate owners for review

**Review by evict — APPROVED** [review]
Looks good from security PoV. :+1:

File: client/web/src/auth/ResetPasswordPage.tsx [hunk 1]
```diff
@@ -164,6 +168,8 @@ class ResetPasswordCodeForm extends React.PureComponent<ResetPasswordCodeFormPro
             <>
                 {isErrorLike(this.state.submitOrError) && <ErrorAlert error={this.state.submitOrError} />}
                 <Container className="w-100">
+                    <Link to='/password-reset'><Icon className="mr-1" aria-hidden={true} svgPath={mdiArrowLeftBoldBoxOutline} />Raise request for a different account</Link>
```
Comment (eseliger) [inline client/web/src/auth/ResetPasswordPage.tsx:171]: Curious about the choise of this icon 😬

File: client/web/src/auth/ResetPasswordPage.tsx [hunk 2]
```diff
@@ -164,6 +168,8 @@ class ResetPasswordCodeForm extends React.PureComponent<ResetPasswordCodeFormPro
...
+                    <Text className="mt-1 text-center text-muted font-weight-bold mb-5">{email}</Text>
```
Comment (eseliger) [inline client/web/src/auth/ResetPasswordPage.tsx:172]: Why the mb-5? I think the regular margin to separate sections would be mb-3 :)

File: cmd/frontend/backend/users.go [hunk 3]
```diff
@@ -26,5 +26,9 @@ func MakePasswordResetURL(ctx context.Context, db database.DB, userID int32) (*u
 	query := url.Values{}
 	query.Set("userID", strconv.Itoa(int(userID)))
 	query.Set("code", resetCode)
+
+	// This field will be used by the frontend for displaying the email on the password entry page
+	query.Set("email", email)
```
Comment (eseliger) [inline cmd/frontend/backend/users.go:31]: @evict is it okay that you can forge a URL that tells you you're resetting your password for a different URL? There's nothing here validating this email is indeed what is being updated.

The safer option would be to request from the server given the PW reset token for what email it was issued.

**Review by eseliger — COMMENTED** [review]
Open question about forging the UI - otherwise LGTM

File: cmd/frontend/backend/users.go (hunk 3 above)
Comment (evict) [reply cmd/frontend/backend/users.go:31]: I don't see any risks coming from this. :thinking: This password reset URL comes into my email, with that field populated. That field only seems to be aesthetic.

**Review by eseliger — APPROVED** [review]
Talked to Vincent he didn't have concerns

File: client/web/src/auth/ResetPasswordPage.tsx (hunk 1 above)
Comment (namit-chandwani) [reply client/web/src/auth/ResetPasswordPage.tsx:171]: I realized that this project uses icons from the `@mdi/js` library. After browsing the available options, I selected the one that seemed to best fit our use case 😅

File: client/web/src/auth/ResetPasswordPage.tsx (hunk 2 above)
Comment (namit-chandwani) [reply client/web/src/auth/ResetPasswordPage.tsx]: Resolved via commit: https://github.com/sourcegraph/sourcegraph/pull/59645/commits/f40722570e30f0aa3d725ce1c96ff36b61389cfc

**Comment by namit-chandwani** [comment]
Hey @eseliger @evict, thanks for the approval 🚀 

Can we please merge the PR if it's good to go?

**Comment by evict** [comment]
> Hey @eseliger @evict, thanks for the approval 🚀
> 
> Can we please merge the PR if it's good to go?

Done! Thank you very much for working on this. :heart_hands:  Apologies that it took a while from our side.

**Comment by namit-chandwani** [comment]
> > Hey @eseliger @evict, thanks for the approval 🚀
> > Can we please merge the PR if it's good to go?
> 
> Done! Thank you very much for working on this. 🫶 Apologies that it took a while from our side.

@evict No worries, I appreciate your time and feedback. Looking forward to contributing more to this project 🙌 
<br>
Furthermore, I have another PR that requires review. Can you please direct me to someone who can assist with it? 
Here's the link for the same: https://github.com/sourcegraph/sourcegraph/pull/60876

TIA

**Pull request merged**